src/target/meson.build

```diff
@@ -0,0 +1,206 @@
+# This file is part of the Black Magic Debug project.
+#
+# Copyright (C) 2023 1BitSquared <info@1bitsquared.com>
+# Written by Rafael Silva <perigoso@riseup.net>
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are met:
+#
+# 1. Redistributions of source code must retain the above copyright notice, this
+#    list of conditions and the following disclaimer.
+#
+# 2. Redistributions in binary form must reproduce the above copyright notice,
+#    this list of conditions and the following disclaimer in the documentation
+#    and/or other materials provided with the distribution.
+#
+# 3. Neither the name of the copyright holder nor the names of its
+#    contributors may be used to endorse or promote products derived from
+#    this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+target_common_includes = include_directories('.')
+
+target_common_sources = files(
+	'adiv5.c',
+	'adiv5_jtag.c',
+	'adiv5_swd.c',
+	'gdb_reg.c',
+	'jtag_devs.c',
+	'jtag_scan.c',
+	'sfdp.c',
+	'spi.c',
+	'target.c',
+	'target_flash.c',
+	'target_probe.c',
+)
+
+# TODO: What is the purpose of this? blindly ported from Makefile
+if get_option('no_own_ll')
+	target_common_sources += files(
+		'jtagtap_generic.c',
+		'swdptap_generic.c',
+	)
+endif
+
+# Conditional file compilation based on target selection
+# We declare a dependency for each target group with the source files required
+# these dependencies are then added to the BMD core, conditinal on the targets option
+# NOTE: sourceset module might be an alternative to this method (unexplored)
```

Comment:
It's certainly an unorthodox use of dependencies, and we'll have to review the Meson manual to verify that it won't break sometime in the future, but assuming it matches documented allowed usage, this works and is fine as it reflects how the KConfig stuff is likely to work and be implemented.

Reply:
Hey! The point of `declare_dependency` is to make them comparable to system-provided libraries, and usable by subprojects. It's an unorthodox usage what @perigoso has done here, but it is needed in order to convey the interdependency between targets.

For porting the FFmpeg project, we instead defined each "target" separately and then described a map of libraries that can be reflected upon as Perigoso did with `get_variable`. But IMO I approve of the way it's been done here, there's no need for such a complicated approach.

Reply:
Perfect, we'll mark this as resolved then as nothing to change 😄

Reply:
It is an unorthodox usage indeed, but it is allowed and should continue to work for us into the future, and I'm fairly happy with it, this is a fairly complex use case and there's no native or recommended approach. As an interesting note, you can still see my initial implementation for target selection on the original PR, where I went with a different approach using dictionaries: https://github.com/perigoso/blackmagic/blob/meson/src/target/meson.build

Also I'll note, I did work closely with the documentation for everything in this PR, I avoided any undefined, deprecated, or unstable API (with the exception of the libopencm3 wrap hack), but double checking doesn't hurt